### DISCUSSION TITLE
Overset mesh for ADflow-MELD-TACS in Mphys

### DISCUSSION THREAD

**Discussion by chlee9000** [OpenMDAO/mphys#179] [Q&A] [answered]
Hello everyone,
May I ask about an error when using overset meshes with ADflow-MELD-TACS in Mphys? 
I'm studying the example in mphys/examples/aerostructural/mach_tutorial_wing. I'm trying to redo the example using a different geometry, .bdf, and an overset CFD mesh. 

I understand that MELD uses nodal _forces_ `f_aero` for loads transfer. So I have `forcesAsTractions=False`.
However, I get this error when I run the aerostructural problem:
```
#* Error message: getForces() is not implmented for zipper meshes and
#*                'forcesAsTractions=False
```
I see that this comes from mphys_adflow.py > ADflowForces.compute. ADflow's getForces.F90 elaborates further: 
```
    if (.not. forcesAsTractions) then
        ! We have a zipper and regular forces are requested. This is not yet supported.
        call terminate('getForces', 'getForces() is not implmented for zipper meshes and '&
             &'forcesAsTractions=False')
    end if
```

Is there some work-around that I'm missing? Or is it simply not currently possible to do an ADflow-MELD-TACS problem so long as near-field meshes overlap on a surface? 

In case it matters, I made wing meshes by following steps in an overset M6 tutorial from the MACH-Aero suite.  For mphys, I specified a user family group: 
`aero_builder = ADflowBuilder(aero_options, scenario="aerostructural", user_family_groups={"wall": ["near_wing", "near_tip"]})`

Thank you!
-Chung

**Comment by anilyil** [comment, marked as answer]
Hi Chung,

I think your understanding is correct. Right now, ADflow cannot output nodal forces on zippers. As a result, if you want to use MELD, you should set up your overset grid so that it does not have any zippers on the wing. However, I think a separate component where we don't transfer the forces (e.g. fuselage) can have zippers. Another way to achieve this would be to turn off zipper meshes, but this would result in wrong force integrations. 

At this stage, I recommend modifying your grid so that there is no wall overlap in the lifting surface. You can still use overset and zippers on the rest of the grid. Please let us know if this does not work. We can try to modify the code so that a zipper mesh thats not on the lifting surface will work with MELD. 

Anil

**Reply by chlee9000** [reply]
Hi Anil,

Thank you for your clear explanation and advice! I'll try such a continuous mesh from winglet to wing-body collar.

If I may ask a follow-up...
One challenge is that a single, continuous grid as described would still be part of an overset system and terminate in a zipper on the fuselage. Not sure whether that would cause the same error in `getForces()`. And, of course, one might simply be interested in strictly the wing surface for aerostructural analysis. So I'm scratching my head on how/whether I'd need to define a subset of wing surface mesh indices for the aerostructures problem. 

I do see Mphys has a `MaskedConverter` component, but not sure how this would work for CFD. I also see in the ADflow builder that there's a `get_tagged_indices`method. But "grepping" in Mphys and ADflow source, it doesn't seem that any other codes or examples are using this for the present purposes. Any thoughts or advice on the need for such masking?

Much appreciated!
- Chung

**Reply by anilyil** [reply]
Hi Chung,

I am not familiar with the get tagged indices method. If I remember correctly, this was implemented to isolate nodes on certain components, e.g. to get the wing nodes from a full wing-body mesh. I have never used it myself so I won't be able to help here. 

I think what should be possible to do is a "half-collar" grid. This would look like a multiblock wing mesh connected to the domains needed on the fuselage for overlap. Then the only zipper mesh would be on the fuselage, which would bypass the requirement. 

Depending on how seriously you are considering using ADflow, it may be a good idea for us to have a quick call. Can you send me an email (the address is available here: https://mdolab.engin.umich.edu/staff_members/ayildirim) to discuss further?

**Reply by A-CGray** [reply]
Hi Chung

Just to add to what Anil said: 

- You can read some more discussion about the details of the zipper mesh force calculation issues in [this ADflow PR](https://github.com/mdolab/adflow/pull/298)
- Using the `get_tagged_indices` approach is the way to go for defining which nodes on the aero and structural meshes are involved in the transfer. However, you don't actually have to call `get_tagged_indices` yourself, you just pass the names of the surfaces you want to be involved in the transfer to the MELD builder when you create it, for example:

```python
# Only transfer loads and displacements between the wall surface of the aero mesh and the skins and spars of the structural mesh
ldTransferBodies = [{"aero": ["wall"], "struct": ["SKIN", "SPAR"]}]

ldxfer_builder = MeldBuilder(
    aero_builder,
    struct_builder,
    isym=isym,
    body_tags=ldTransferBodies,
)
```

**Reply by chlee9000** [reply]
Thank you for your replies, Anil and Alasdair!
The PR and your explanations are very informative. I'll study these carefully and try to apply them along with the "half collar" approach (and also ask Anil off-line about some details).
Thanks again!
-Chung
